subprojects/testing-junit-platform/src/main/java/org/gradle/api/internal/tasks/testing/junitplatform/JUnitPlatformTestExecutionListener.java

```diff
@@ -0,0 +1,218 @@
+        currentRunningTestClass.end(className(testIdentifier));
+    }
+
+    private boolean isLeafMethodOrFailedContainer(TestIdentifier testIdentifier, TestExecutionResult result) {
+        // Generally, there're 2 kinds of identifier:
+        // 1. A container (test engine/class/repeated tests). It is not tracked unless it fails/aborts.
+        // 2. A test "leaf" method. It's always tracked.
+        return isLeafTest(testIdentifier) || isFailedContainer(testIdentifier, result);
+    }
+
+    private boolean isFailedContainer(TestIdentifier testIdentifier, TestExecutionResult result) {
+        return result.getStatus() != SUCCESSFUL && testIdentifier.isContainer();
+    }
+
+    private TestDescriptorInternal getDescriptor(final TestIdentifier test) {
+        if (isMethod(test)) {
+            return new DefaultTestDescriptor(idGenerator.generateId(), className(test), test.getDisplayName());
+        } else if (isVintageDynamicTest(test)) {
+            UniqueId uniqueId = UniqueId.parse(test.getUniqueId());
+            return new DefaultTestDescriptor(idGenerator.generateId(), vintageDynamicClassName(uniqueId), vintageDynamicMethodName(uniqueId));
+        } else if (isVintageNullDescriptionTest(test)) {
+            return new DefaultTestDescriptor(idGenerator.generateId(), className(test), test.getDisplayName());
+        } else if (isClass(test)) {
+            return new DefaultTestDescriptor(idGenerator.generateId(), className(test), "classMethod");
+        } else {
+            return null;
+        }
+    }
+
+    private boolean isMethod(TestIdentifier test) {
+        return test.getSource().isPresent() && test.getSource().get() instanceof MethodSource;
+    }
+
+    private boolean isClass(TestIdentifier test) {
+        return test.getSource().isPresent() && test.getSource().get() instanceof ClassSource;
+    }
+
+    private String className(TestIdentifier testIdentifier) {
+        if (isMethod(testIdentifier)) {
+            return retrieveClassName(testIdentifier);
+        } else if (testIdentifier.getSource().get() instanceof ClassSource) {
+            return ClassSource.class.cast(testIdentifier.getSource().get()).getClassName();
+        } else {
+            return "UnknownClass";
+        }
+    }
+
+    private String retrieveClassName(TestIdentifier testIdentifier) {
+        // For tests in default method of interface,
+        // we might not be able to get the implementation class directly.
+        // In this case, we need to retrieve test plan to get the real implementation class.
+        TestIdentifier current = testIdentifier;
+        while (!isClass(current) && current.getParentId().isPresent()) {
+            current = currentTestPlan.getTestIdentifier(current.getParentId().get());
+        }
+
+        return className(current);
+    }
+
+    private class CurrentRunningTestClass {
+        private String name;
```

Comment:
We use this class to track class start/end event. Unlike previous `TestClassProcessor` which processes classes one by one, now we send all test classes in one pass to `JUnit Platform`. This means JUnit Platform will generate a complex test plan where test class start/end event can be triggered more than one time. With `CurrentRunningTestClass`, we can exactly know when a test class starts and ends.

Reply:
This will likely break as soon as JUnit Jupiter/Platform introduces parallel test execution.

Reply:
Yep, but unfortunately, current Gradle infrastructure assume all tests are executed serially in a single worker process so that all events can be tracked correctly. Do you have any design doc about the upcoming parallel functionality. I guess we need to turn off parallel execution at the beginning.